test/methods/gd_lipschitz_approximation.jl

```diff
@@ -0,0 +1,172 @@
+        )
+
+        for field_elem in fields
+            #Test Names (does not depend on field_type)
+            @test field_elem[1] in fieldnames(LipschitzApproxGD)
+
+            #Test Types
+            @test field_elem[2] == typeof(getfield(optData, field_elem[1]))
+        end
+    end
+
+    ## Test Assertions
+    @test_throws AssertionError LipschitzApproxGD(
+        Float64,
+        x0 = randn(param_dim),
+        init_stepsize = 0.0, 
+        threshold = 1e-4,
+        max_iterations = 100
+    )
+
+    ##########################################
+    # Test optimizer
+    ##########################################
+
+    progData = OptimizationMethods.GaussianLeastSquares(Float64)
+    x0 = progData.meta.x0
+    ss = 1e-2
+
+    ## One Step Update 
+    optData = LipschitzApproxGD(
+        Float64,
+        x0 = x0,
+        init_stepsize = ss,
+        threshold = 1e-10,
+        max_iterations = 1
+    )
+
+    x1 = lipschitz_approximation_gd(optData, progData)
+
+    ### Test updated values 
+    g0 = OptimizationMethods.grad(progData, x0)
+    g1 = OptimizationMethods.grad(progData, x1)
+    @test x1 ≈ x0 - ss * g0
+
+    ### Test Stored values
+    @test optData.prev_stepsize == ss
+    @test optData.theta == Inf64
+    @test optData.lipschitz_approximation ≈ norm(g1 - g0) / norm(x1 - x0)
+    @test optData.iter_diff ≈ x1 - x0
+    @test optData.grad_diff ≈ g1 - g0
+    @test optData.iter_hist[1] == x0
+    @test optData.iter_hist[2] == x1
+    @test optData.grad_val_hist[1] ≈ norm(g0)
+    @test optData.grad_val_hist[2] ≈ norm(g1)
+    @test optData.stop_iteration == 1
+
+
+    ## Two Step Update 
+    optData = LipschitzApproxGD(
+        Float64,
+        x0 = x0,
+        init_stepsize = ss,
+        threshold = 1e-10,
+        max_iterations = 2
+    )
+
+    x2 = lipschitz_approximation_gd(optData, progData)
+    
+    ### Test updated values 
+    g0 = OptimizationMethods.grad(progData, x0)
+    x1 = optData.iter_hist[2]
+    g1 = OptimizationMethods.grad(progData, x1)
+    g2 = OptimizationMethods.grad(progData, x2)
+    @test x2 ≈ x1 - (norm(x1 - x0))/(2*norm(g1-g0)) * g1
+
+    ### Test Stored values
+    @test optData.prev_stepsize ≈ (norm(x1 - x0))/(2*norm(g1-g0))
+    @test optData.theta ≈ optData.prev_stepsize / ss
+    @test optData.lipschitz_approximation ≈ norm(g2 - g1) / norm(x2 - x1)
+    @test optData.iter_diff ≈ x2 - x1
+    @test optData.grad_diff ≈ g2 - g1
+    @test optData.iter_hist[3] == x2 
+    @test optData.grad_val_hist[2] ≈ norm(g1)
+    @test optData.grad_val_hist[3] ≈ norm(g2)
+    @test optData.stop_iteration == 2
+
+
+    ## K Step Update K == 52
+    optData = LipschitzApproxGD(
+        Float64,
+        x0 = x0,
+        init_stepsize = ss,
+        threshold = 1e-10,
+        max_iterations = 100
+    )
+
+    xk = lipschitz_approximation_gd(optData, progData)
+
+    ### Test updated values
+    k = optData.stop_iteration
+    xkm1 = optData.iter_hist[k-1]
```

Comment:
I think all these indices should be incremented by 1 since `xk` is stored at `k+1`.

Reply:
This comment is still outstanding (if you change the max iteration to 7, all these test will fail).